docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+  - `__llvm_bitcast_pi64_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi64_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi64_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi64_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi64_to_pi128 -> pi128`
+  - `__llvm_bitcast_pi128_to_pbool -> pbool`
+  - `__llvm_bitcast_pi128_to_pi8 -> pi8`
+  - `__llvm_bitcast_pi128_to_pi16 -> pi16`
+  - `__llvm_bitcast_pi128_to_pi32 -> pi32`
+  - `__llvm_bitcast_pi128_to_pi64 -> pi64`
+  - `__llvm_bitcast_pi128_to_pi128 -> pi128`
+- [`icmp`](https://llvm.org/docs/LangRef.html#icmp-instruction):
+  - `icmp` accepts three arguments. The first is a comparison condition. The two others are operands
+    of the same type. The condition is defined as an enum consisting of these values: `eq`, `ne`,
+    `ugt`, `uge`, `ult`, `ule`, `sgt`, `sge`, `slt`, `sle`.
+  - `__llvm_icmp_cond_bool_bool -> bool`,
+  - `__llvm_icmp_cond_i8_i8 -> bool`,
+  - `__llvm_icmp_cond_i16_i16 -> bool`,
+  - `__llvm_icmp_cond_i32_i32 -> bool`,
+  - `__llvm_icmp_cond_i64_i64 -> bool`,
+  - `__llvm_icmp_cond_i128_i128 -> bool`,
+- [`select`](https://llvm.org/docs/LangRef.html#select-instruction):
+  - `__llvm_select_bool_bool_bool -> bool`,
+  - `__llvm_select_bool_i8_i8 -> bool`,
+  - `__llvm_select_bool_i16_i16 -> bool`,
+  - `__llvm_select_bool_i32_i32 -> bool`,
+  - `__llvm_select_bool_i64_i64 -> bool`,
+  - `__llvm_select_bool_i128_i128 -> bool`,
+
+#### Based on intrinsics
+
+- [`llvm.abs.*`](https://llvm.org/docs/LangRef.html#llvm-abs-intrinsic):
+  - `__llvm_abs_i8 -> i8`,
+  - `__llvm_abs_i16 -> i16`,
+  - `__llvm_abs_i32 -> i32`,
+  - `__llvm_abs_i64 -> i64`,
+  - `__llvm_abs_i128 -> i128`,
+- [`llvm.smax.*`](https://llvm.org/docs/LangRef.html#llvm-smax-intrinsic):
+  - `__llvm_smax_bool_bool -> bool`,
+  - `__llvm_smax_i8_i8 -> i8`,
+  - `__llvm_smax_i16_i8 -> i16`,
+  - `__llvm_smax_i32_i8 -> i32`,
+  - `__llvm_smax_i64_i8 -> i64`,
+  - `__llvm_smax_i128_i8 -> i128`,
+- [`llvm.smin.*`](https://llvm.org/docs/LangRef.html#llvm-smin-intrinsic):
+  - `__llvm_smin_bool_bool -> bool`,
+  - `__llvm_smin_i8_i8 -> i8`,
+  - `__llvm_smin_i16_i8 -> i16`,
+  - `__llvm_smin_i32_i8 -> i32`,
+  - `__llvm_smin_i64_i8 -> i64`,
+  - `__llvm_smin_i128_i8 -> i128`,
+- [`llvm_umax.*`](https://llvm.org/docs/LangRef.html#llvm-umax-intrinsic):
+  - `__llvm_umax_bool_bool -> bool`,
+  - `__llvm_umax_i8_i8 -> i8`,
+  - `__llvm_umax_i16_i8 -> i16`,
+  - `__llvm_umax_i32_i8 -> i32`,
+  - `__llvm_umax_i64_i8 -> i64`,
+  - `__llvm_umax_i128_i8 -> i128`,
+- [`llvm.umin.*`](https://llvm.org/docs/LangRef.html#llvm-umin-intrinsic):
+  - `__llvm_umin_bool_bool -> bool`,
+  - `__llvm_umin_i8_i8 -> i8`,
+  - `__llvm_umin_i16_i8 -> i16`,
+  - `__llvm_umin_i32_i8 -> i32`,
+  - `__llvm_umin_i64_i8 -> i64`,
+  - `__llvm_umin_i128_i8 -> i128`,
+- [`llvm.scmp.*`](https://llvm.org/docs/LangRef.html#llvm-scmp-intrinsic):
+  - `scmp` returns needs to return at least `i2`. Since ALU does not operate on such type, the
+    closest possible type is `i8`.
```

Comment:
We need to work out what is going to be seen in LLVM IR. The compiler is not going to be built to handle type extension like this, so if we see instructions returning `i2` we probably need to support it.

Reply:
I am not suggesting that _we_ support all $2^{32}$ integer types, just that we need to find out what `rustc` generates in LLVM IR for the return type. If it generates `i8` we're golden, but if it generates `i2` it is beyond the scope of the compiler to handle that and we would need to provide appropriate polyfill operations.